marlowe-dashboard-client/src/MainFrame/State.purs

```diff
@@ -45,81 +48,138 @@ mkMainFrame =
+
+handleAction PickupNewWallet = do
+  newPubKeyHash <- use (_newWalletNicknameKey <<< _key)
+  handleAction AddNewWallet
+  handleAction $ PickupWallet newPubKeyHash
+
+handleAction (LookupWallet string) = do
+  wallets <- use _wallets
+  for_ (findIndex (\key -> fst key == string) wallets) $ \key -> assign (_outsideState <<< _card) $ Just $ PickupWalletCard key
+
+handleAction (PickupWallet pubKeyHash) = do
+  modify_
+    $ (set _subState) (Right $ mkInsideState pubKeyHash)
+    <<< (_outsideState <<< set _card) Nothing
+  liftEffect $ setItem walletLocalStorageKey $ encodeJSON pubKeyHash
+
+-- inside actions
+handleAction PutdownWallet = do
+  assign _subState $ Left initialOutsideState
+  liftEffect $ removeItem walletLocalStorageKey
+
+handleAction ToggleMenu = modifying (_insideState <<< _menuOpen) not
 
 handleAction (SetScreen screen) =
   modify_
-    ( set _overlay Nothing
-        <<< set _card Nothing
-        <<< set _screen screen
-    )
+    $ (_insideState <<< set _menuOpen) false
+    <<< (_insideState <<< set _card) Nothing
+    <<< (_insideState <<< set _screen) screen
+
+handleAction (SetCard card) = do
+  previousCard <- peruse (_insideState <<< _card)
+  assign (_insideState <<< _card) card
+  for_ previousCard $ const $ assign (_insideState <<< _menuOpen) false
 
 handleAction (ToggleCard card) = do
-  assign _overlay Nothing
-  mCurrentCard <- use _card
+  mCurrentCard <- peruse (_insideState <<< _card)
   case mCurrentCard of
     Just currentCard
-      | card == currentCard -> assign _card Nothing
-    _ -> assign _card $ Just card
+      | currentCard == Just card -> handleAction $ SetCard Nothing
+    _ -> handleAction $ SetCard $ Just card
 
-handleAction CloseCard = assign _card Nothing
+handleAction (SetNewWalletNickname nickname) = assign (_newWalletNicknameKey <<< _nickname) nickname
 
-handleAction (ContactAction contactAction) = do
-  mapSubmodule _contactState ContactAction $ Contact.handleAction contactAction
-  case contactAction of
-    Contact.ToggleNewContactCard -> handleAction $ ToggleCard NewContact
-    Contact.AddNewContact -> handleAction $ ToggleCard NewContact
-    Contact.ToggleEditContactCard contactKey -> handleAction $ ToggleCard $ EditContact contactKey
-    _ -> pure unit
+handleAction (SetNewWalletKey key) = assign (_newWalletNicknameKey <<< _key) key
+
+handleAction AddNewWallet = do
+  oldWallets <- use _wallets
+  newWalletNicknameKey <- use _newWalletNicknameKey
+  when (not $ member newWalletNicknameKey oldWallets) do
+    modifying _wallets $ insert newWalletNicknameKey defaultWalletDetails
+    assign _newWalletNicknameKey emptyWalletNicknameKey
+    newWallets <- use _wallets
+    liftEffect $ setItem walletsLocalStorageKey $ encodeJSON newWallets
+    assign (_insideState <<< _card) Nothing
+
+handleAction ClickedButton = do
+  mCurrent <- peruse (_insideState <<< _on)
+  for_ mCurrent $ \current -> raise (SendWebSocketMessage (ServerMsg current))
```

Comment:
I wish there was a version of `for_` that was `specialfor_ :: Applicative m => Foldable f => m (f a) -> (a -> m b) -> m Unit` so we could have `specialfor_ (peruse (insideState <<< _on)) \current -> ...`